WIP Crs features ogc

justb4 opened the pull request:
# Overview
See PRs 1155 and #1149. 
This PR combines these two as together they provide support
for the complete standard [OGC API - Features - Part 2: Coordinate Reference Systems by Reference](https://docs.opengeospatial.org/is/18-058r1/18-058r1.html).

# Related Issue / Discussion
Issue #1128 - See PRs 1155 and #1149. 

# Additional Information
See related PRs above.

# Contributions and Licensing

(as per https://github.com/geopython/pygeoapi/blob/master/CONTRIBUTING.md#contributions-and-licensing)

- [x] I'd like to contribute this feature  to pygeoapi. I confirm that my contributions to pygeoapi will be compatible with the pygeoapi license guidelines at the time of contribution.
- [x] I have already previously agreed to the pygeoapi Contributions and Licensing Guidelines

justb4 commented:
The CI is failing from what I think is a Proj/GDAL version issue. The test that fails is the OGR Provider which uses internal GDAL calls for the Transform (i.s.o. using Shapely/pyproj as in api.py) when a CrsTransform is passed on get() or query().

On my local system  this unit test succeeds with versions:
```
GDAL 3.2.2, released 2021/03/05
pyproj==3.2.1
libproj 7.2.1, January 1st, 2021
```

The CI uses Ubuntu 20.04 without UbuntuGIS so may have different versions of these.

justb4 commented:
Probably best is to upgrade `main.yml` from Ubuntu 20.04 to 22.04. This is also inline with the Docker Image. As a start.
Test runs fine in Docker Image. I opened #1176 as it has quite some impact.

```
platform linux -- Python 3.10.6, pytest-6.2.5, py-1.10.0, pluggy-0.13.0
rootdir: /pygeoapi, configfile: pytest.ini
collected 17 items

tests/test_ogr_shapefile_provider.py .................                   [100%]
```

But interesting, 2 errors in the WFS tests with the GeoSolutions WFS:

```
platform linux -- Python 3.10.6, pytest-6.2.5, py-1.10.0, pluggy-0.13.0
rootdir: /pygeoapi, configfile: pytest.ini
collected 17 items

tests/test_ogr_shapefile_provider.py .................                   [100%]
```

MTachon commented:
I agree with @justb4 , that the axis ordering in GeoJSON responses should follow the definition of the CRS, and not forced to lon/lat for geographic coordinate systems.

The section [Coordinate reference system information independent of the feature encoding](https://docs.opengeospatial.org/is/18-058r1/18-058r1.html#_coordinate_reference_system_information_independent_of_the_feature_encoding) introduce the `Content-Crs` header as a solution for *"...the continued confusion caused by the axis order of coordinates, ..."*. It defines it as *"a mechanism for a server to clearly and unambiguously assert the CRS being used in a response document independent of the requested output format."*. "Natively", GeoJSON does not support other CRSs than WGS84 lon/lat (OGC:CRS84), but the [Part 2 of the standard OGC API - Features](https://docs.opengeospatial.org/is/18-058r1/18-058r1.html) mentions the *"the prior arrangement provision [in the second paragraph of section 4 of the GeoJSON standard](https://www.rfc-editor.org/rfc/rfc7946.html#section-4)"*, which *"allows other coordinate systems to be used"*. It also states: *"An explicit request by a client with a query parameter `crs` establishes a prior arrangement. It is the responsibility of the client that submits the request to handle the coordinates in the response correctly. "*. By *handling the coordinates correctly*, I would assume that axis ordering of the features' coordinates in the response should follow that of the queried CRS.

As for a *solution*, pyproj can perfectly handle the axis ordering for different CRSs with the `always_xy` parameter:

```python

# Input CRS with lat/lon axis order
crs_in = pyproj.CRS.from_espg(4326)

# Output CRS with lon/lat axis order
crs_out = pyproj.CRS.from_authority("OGC", "CRS84")

# Create a pyproj.Transformer instance which forces lon/lat ordering
transformer_1 = pyproj.Transformer.from_crs(crs_in, crs_out, always_xy=True)

# Create a pyproj.Transformer instance which takes care of changing axis order if needed
transformer_2 = pyproj.Transformer.from_crs(crs_in, crs_out, always_xy=False)

>>> transformer_1.transform(5.0, 6.0)
(5.0, 6.0)

>>> transformer.transform(5.0, 6.0)
(6.0, 5.0)
```

On the other hand, when looking at the implementation of the `OGRProvider`, I see that it is using the `SetAxisMappingStrategy` method of the `OGRSpatialReference` to set the axis order to `osr.OAMS_TRADITIONAL_GIS_ORDER`, which forces the lon/lat axis order for coordinates transformation. In other words, the transformation will *expect* coordinates in lon/lat order as input (independently of the data source and its support for different coordinate axis order for different CRSs), and *spit out* coordinates in the same order. As I understand it, the GDAL/OGR `OGRSpatialReference.SetAxisMappingStrategy(osr.OAMS_TRADITIONAL_GIS_ORDER)` is the same as the pyproj `always_xy=True`. If we decide to change one of them, the other should be changed accordingly.

As for data stored in PostGIS, it looks like the `ST_Transform` function switches the coordinates axis order if needed, and follows the CRS axis order convention.

MTachon commented:
Something else that came to my mind when reading the [Part 2 of the standard OGC API - Features](https://docs.opengeospatial.org/is/18-058r1/18-058r1.html), is the implementation of pygeoapi when serving data in JSON for Linking Data format (JSON-LD). Currently, the `linked_data.geojson2jsonld` function uses the X/Y ordering, stored in shapely geometrical object. Coordinates order should be switched in this function, as coordinates in shema.org markup (`GeoShape`/`GeoCoordinates`) only supports WGS 84 in the axis order latitude/longitude. Part 2 of the standard also says *"...any coordinates in schema.org markup will have to be in that coordinate reference system, independent of the requested coordinate reference system."*. Depending on which `crs` is queried, or which `storage_crs` is used, if the user chooses the JSON-LD data format, the coordinates in the schema.org markup may need be transformed to `EPSG:4326`...

justb4 commented:
@MTachon wrote:
> If we decide to follow @justb4 's convention, setting always_xy=False in util.get_transform_from_crs and > util.transform_bbox, as well as removing the occurences of > OGRSpatialReference.SetAxisMappingStrategy(osr.OAMS_TRADITIONAL_GIS_ORDER) in the OGRProvider may be enough?

Yes, this is my current plan for this PR/branch, and get all tests working, maybe add a few more. Like with BBOX-CRS I have several to check for CRS-compliant Axis ordering.

MTachon commented:
I do not think that it has something to do with the projection being loaded from WKT, but more with `osr.OAMS_AUTHORITY_COMPLIANT` vs. `osr.OAMS_TRADITIONAL_GIS_ORDER` in the `_get_spatial_reference()`. See my comment for `ogr.py` file.

justb4 commented:
> I do not think that it has something to do with the projection being loaded from WKT, but more with `osr.OAMS_AUTHORITY_COMPLIANT` vs. `osr.OAMS_TRADITIONAL_GIS_ORDER` in the `_get_spatial_reference()`. See my comment for `ogr.py` file.

Have you seen the latest changes in `ogr.py`, on line 79:
CRS_AXIS_ORDER = osgeo_osr.OAMS_AUTHORITY_COMPLIANT
This setting is used for all Projection loading, EPSG (except 4326) used with the source/target EPSG config and WKT as for the Transformation object passed in `get()` and `query()`. Still the Projection/Transformation using the latter WKT loading does not honour CRS Axis ordering, e.g. for 4258. For now I have disabled Transformation for `ogr.py` and use `@crs_transform` until sorted out. I think the WKT loading also caused the CI `main.yml` to fail (vague error pointing to 'WKT1 2019' or something). Unless someone can fix this, I propose to open a new issue/PR. Maybe the CrsTransformation Object can also specify/embed two EPSG code-`ints`, such that `ogr.py` does loading from EPSG i.s.o. WKT as the latter seems 'brittle' or maybe even the full Projection URIs, maybe that works as well. Other Providers may be helped with this as well.

Only 'gotcha' now may be `storageCRS` which should be set now to the 'target' EPSG, as in some cases there may be two reprojections.

MTachon commented:
So if I understand correctly, importing a CRS (into a `OGRSpatialReference` object) from WKT gives a different result than when importing a CRS from EPSG? Or do you get an error when mixing `source_srs`/`target_srs` and `crs`/`storage_crs`? What dataset are you using for testing (and which CRS and axis ordering does it have)? What it the output CRS wanted? The CI `main.yml` may be related to a WKT version issue, resulting from pyproj's PROJ version vs GDAL's PROJ version. Specifying the WKT version can be done with pyproj when exporting CRSs (and constructing the `CrsTransformWkt` instance). If changing this works, it may be an easier fix than fixing `main.yml`.

justb4 commented:
> So if I understand correctly, importing a CRS (into a `OGRSpatialReference` object) from WKT gives a different result than when importing a CRS from EPSG? Or do you get an error when mixing `source_srs`/`target_srs` and `crs`/`storage_crs`? 

Yes,  that looks like it: importing from WKT gives either wrong (Axis ordering) result in Docker deploy or failure in `main.yml` CI. Also when disabling `source_srs`/`target_srs`. Happens with any `storage_crs`: http://www.opengis.net/def/crs/OGC/1.3/CRS84 (stored in lon,lat) or http://www.opengis.net/def/crs/EPSG/0/28992 (always x,y).

> What dataset are you using for testing (and which CRS and axis ordering does it have)? What is the output CRS wanted? The CI `main.yml` may be related to a WKT version issue, resulting from different pyproj's PROJ and GDAL's PROJ versions. Specifying the WKT version can be done with pyproj when exporting CRSs (and constructing the `CrsTransformWkt` instance). If changing this works, it may be an easier fix than fixing `main.yml`.

It follows from the unit tests. Using OGR with GPKG with several projections. You can see the testfile/URL here:
https://apitestbed.geonovum.nl/test/ . All CRS, BBOX-CRS tests here work as expected now, as the OGR Provider delegates transforms to core (via `@crs_transform` decorator). The pygeoapi instance runs from the repo branch `crs-features-ogc-adr` as Docker instance. Find data and config here.

My personal opinion is that WKT loading seems brittle among pyproj, Proj, GDAL. Users will run different versions. Will be hard to grasp this. That is why I suggested to to pass a `Transform` object with also the original EPSG code or maybe URI to `get()` and `query()`. At least within OGR Driver internal Projection instantiation is uniform, both via `source/target` and `crs` parameter, i,e, load from EPSG with `osgeo_osr.OAMS_AUTHORITY_COMPLIANT`.

justb4 commented:
Now running the official OGC "Cite" tests, for that instance, interesting results. Should have run earlier, very useful for development, easy to get running with Docker: https://cite.opengeospatial.org/teamengine/about/ogcapi-features-1.0/1.0/site/.
Aiming now for zero fails for Part 2...

justb4 commented:
@MTachon Hmm, I had hoped a bit for the latter `OGRSpatialReference.ImportFromUrl()`. On the other hand  http://www.opengis.net/def/crs/OGC/1.3/CRS84 is the default, basically  http://www.opengis.net/def/crs/EPSG/0/4326 with lon,lat ordering so maybe in that case use `OAMS_TRADITIONAL_GIS_ORDER`. 

As we have quite some tests and the CITE Tester running I propose that I:

* extend the `CrsTransformWkt` with the original URIs
* use `OGRSpatialReference.ImportFromUrl()` in the OGR Provider (with `OAMS_TRADITIONAL_GIS_ORDER` for http://www.opengis.net/def/crs/OGC/1.3/CRS84 otherwise `OAMS_AUTHORITY_COMPLIANT`.

btw I also was a bit surprised to see that the PostgreSQL Provider uses the "Shapely-based-Transform". I would expect that transformation would be delegated to PostGIS `ST_Transform()`, but ok, later versions of the Provider now use SQLAlchemy plus `ST_Transform()` would possibly get us into another projection mess, as the Proj info is in the table `spatial_ref_sys`...and is all SRID (int) based AFAICS.

MTachon commented:
Have had much trouble with network in new offices, have not had time to check the CITE tester.

The reason for using the "Shapely-based-Transform" in the `PostgreSQLProvider` is that it uses sqlalchemy/geoalchemy to query the entire table in the database (`session.query(self.table_model)`). To use spatial functions (e.g. `ST_Transform()`) with geoalchemy2, one must query the geometry/geography column(s). The other columns of the queried table (Features' properties) are not to be transformed. Using geoalchemy2 is impractical for our case, as it would require multiple queries (or maybe subqueries), one for geometries, one for properties, to get the features with transformed geometries and untouched properties. As the `PostgreSQLProvider` already used shapely to construct geometrical objects from their (E-)WKT/WKB representations, as returned by geoalchemy2, before being exported to GeoJSON, it seemed to be the most straightforward approach, and it does not really add overhead either (PostGIS-PROJ transforms vs pyproj-PROJ transforms). If we absolutely wanted to use PostGIS's backend (`ST_Transform()` function), I think it would be easier to build the SQL query ourselves than using the ORM approach.

justb4 commented:
Ok, I understand. Think the [very first versions of the PostgreSQL Provider used direct SQL](https://github.com/geopython/pygeoapi/blob/e4a36e7248d6c041f840196d1716d85cc705d467/pygeoapi/provider/postgresql.py) (and `ST_ToGeoJSON()`), no sql/geoalchemy, so your approach makes sense here.

justb4 commented:
Well even the URI loading in OGR gave failures in cases. So moved to one uniform Projection loading all over the `OGRProvider`: from EPSG (code), forcing lon,lat where needed. Added many more tests for alle kinds of (CRS transform axis ordering) cases. Up to now all succeeding. Note that I changed `CrsTransformWkt` to `CrsTransformSpec`, adding source/target CRS URIs.

And the more I am thinking: `source/target_srs` only adds confusion and complexity. 
Now thinking of removing. 'Confusing' in the sense:

* if `target_srs` is not http://www.opengis.net/def/crs/OGC/1.3/CRS84 and no `crs` parameter is passed,  a GeoJSON response in that SRS without even a `Content-Crs` header is returned. This is non-conformant.
* it adds/requires an enormous extra amount of test-cases, configurations
* we mix the EPSG-form (SRS) and URI form (CRS)
* combined with `storageCRS` gives complex interpretations
* once Transactions are supported it will get even more complex.
* it is not inline with the other backend Providers
* most important: there is `storageCRS` and `crs` support now, let's use that and stay within OGC Compliance

Blame on me: I had then introduced `source/target_srs`....

Interesting cases remain for any Provider: 

**if the `storage_crs` is not the default http://www.opengis.net/def/crs/OGC/1.3/CRS84(h) AND no `crs` parameter is passed:**

* should Features be returned with `storage_crs` coordinates (and `Content-Crs` header)? 
* or (my pref) in standard GeoJSON lon, lat WGS84 with http://www.opengis.net/def/crs/OGC/1.3/CRS84 in `Content-Crs`

Next question then, who should handle/enforce coord Transformation?

* `api.py` : creating and passing a `CrsTransformSpec` to the Provider? (My pref)
*  each Provider take care of returning proper GeoJSON WGS84, checking `storageCRS`?

MTachon commented:
I agree that having both `source_srs`/`target_srs` and `storage_crs`/`crs`,  as well as mixing EPSG codes and CRS URIs, can be confusing, especially for new users/service administrators. `source_srs`/`target_srs` was certainly a nice feature when introduced, and gave `OGRProvider` some flexibility that other providers did not have. The support for the *CRS* part of the OGC API Feature standard gives even **more flexibility**, to **all providers**, and all providers would **behave the same way**.

Removing it would break some functionality of deployed services for which their configuration file uses `source_srs`/`target_srs` (when the administrators choose to update pygeoapi). It would mean that the data served to end-users would no longer be transformed, maybe the end-users got used to the data being served in the same CRS (`target_srs`)... If we choose to remove it, it has to be well communicated:

- documentation updated, maybe a `.. note::` or deprecation `.. warning::` in the section devoted to publishing data with the `OGRProvider`.
- maybe throwing a warning when starting services that uses the `OGRProvider` and where `source_srs`/`target_srs`  is found in the config file? This warning could be removed after some time, when we think pygeoapi's users are no longer using this feature.

The administrators can then transform their datasets so that the *new* `storage_crs` matches the *old* `target_srs` so that end-users do not get "surprised". Maybe the administrators will need to duplicate their datasets for a time while the use of the `crs` query parameter gets broadly accepted: it is not unlikely that pygeoapi users deployed several services for the same dataset (same `source_srs`), but with different output CRSs (different `target_srs`).

**If the `storage_crs` is not the default http://www.opengis.net/def/crs/OGC/1.3/CRS84(h) AND no crs parameter is passed:**
The **Requirement 12 (/req/crs/fc-crs-default-value)** in the [Parameter crs section](https://docs.opengeospatial.org/is/18-058r1/18-058r1.html#_parameter_crs) states:

*If the crs parameter is not specified the geometry coordinates SHALL be presented in the default CRS specified in [OGC API - Feature - Part 1: Core](https://docs.opengeospatial.org/is/18-058r1/18-058r1.html#OAFeat-1); that is http://www.opengis.net/def/crs/OGC/1.3/CRS84 for coordinates without height and http://www.opengis.net/def/crs/OGC/0/CRS84h for coordinates with ellipsoidal height.*

Returning features in standard GeoJSON lon, lat WGS84 with http://www.opengis.net/def/crs/OGC/1.3/CRS84 in `Content-Crs` would make sense. This means that if `storage_crs` differs from http://www.opengis.net/def/crs/OGC/1.3/CRS84, the data should be transformed (`storage_crs` -> `OGC:CRS84`).

For your last question, I agree with you, that creating and passing a `CrsTransformSpec` to the Provider (in the `api.py` side) is the right approach. I think the decorator approach is fine for providers that do not support natively CRS transformation. `storageCRS` can be checked once (in `api.py`), no need for all providers to extract this information from config. And if neither `storageCRS` nor `crs` are WGS84, the second approach would result in all providers transforming the data once (from `storageCRS` to proper GeoJSON WGS84), and the data would need to be transformed once again in `api.py`(from proper GeoJSON WGS84 to `crs`).

justb4 commented:
@MTachon think we agree on all points! Yes, removing `source/target_crs` may be impacting installations. Indeed issueing a warning is planned, and documentation. Thinking of a transition period, warning that behaviour when combined with `crs` and `bbox-crs` is undefined, or that `WktTransformSpec` is ignored when `target_srs` is set. `source_srs` is I think taken care of with `storageCRS` and upcoming changes in `api.py`, see next. 

Same (as `crs` param) goes for `bbox-crs`, already tested. When the `bbox` param has no `bbox-crs` the default (WGS84, lon/lat) is assumed. When the `storageCRS` is not that default CRS, also a similar Transform for the `bbox` arg to the Provider is needed/done (in `api.py`). This is basically the behaviour of the current `source_srs`.

justb4 commented:
Had to fix CSV and Postgresql tests as well as the default behaviour is not anymore to return Features with storageCRS CRSs but is OGC Compliant WGS84, lon,lat.
All ok IMO.

But the biggest headache now is that in some cases GDAL OGR Transforms do not respect CRS (lat,lon) Axis Ordering for some CRS, namely  http://www.opengis.net/def/crs/EPSG/0/4326 and http://www.opengis.net/def/crs/EPSG/0/4258. Even when creating `SpatialReference` like for example:

```
        spatial_ref = osgeo_osr.SpatialReference()
        spatial_ref.SetAxisMappingStrategy(osgeo_osr.OAMS_AUTHORITY_COMPLIANT)
        spatial_ref.ImportFromEPSG(4258)
``` 
Many tests confirm this, also when using  4326 (lat,lon) as in http://www.opengis.net/def/crs/EPSG/0/4326 . For http://www.opengis.net/def/crs/OGC/1.3/CRS84 '4326' is used with `osgeo_osr.OAMS_TRADITIONAL_GIS_ORDER`. 

The only context where the problem exists (Axis order still lon, lat) is in the Docker Image. It works on my local setup and in the CI `main.yml` (as seen in Actions). Difference are GDAL and Proj versions (and their Python bindings). The Docker Image, unfortunately, has the highest versions:

* My Local Env: Homebrew GDAL 3.2.2 - Proj 7.2.1 - Pyproj 3.0.0 - GDAL Python bindings 3.2.0 (**OK**)
* CI main.yml: Ubuntu 20.04 - GDAL 3.0.4 - Proj 6.3.1-1 - Pyproj ? (comes with Proj?)  - GDAL Python bindings 3.0.4 (**OK**)
* Docker Image: Ubuntu 22.04 - GDAL 3.4.1 - Proj 8.2.1-1 - Pyproj 3.3.0 - GDAL Python bindings 3.4.1 (**not OK**)

At least we have a situation where only in one way each OGR `SpatialReference` is created in `ogr.py`. But it may be that we need to again disable transforms in OGRProvider unless someone can give some insight and a solution. We may leave for example the native OGR Transform in a new issue for sake of progress....(my pref).

Who can shed light @rouault maybe as last resort?

rouault commented:
> Who can shed light @rouault maybe as last resort?

not a lot given that I only have partial context on what you do. Normally all GDAL >= 3 + PROJ >= 6 versions should behave the same for basic CRS such as EPSG:4326 or EPSG:4258 regarding axis order. If you can isolate your issue to a standalone reproducer just involving GDAL API, that would be helpful

justb4 commented:
Thanks @rouault ! This put me on an interesting journey. I isolated the Transformation handling into a unit test. I executed that test in all 3 environments (local, CI and Docker, see above). But even in the Docker Image it succeeded! i.e. Axis order was preserved in all cases using the OGR SpatialReference and Coord Transformations. 

See [the unit test file](https://github.com/geopython/pygeoapi/blob/crs-features-ogc/tests/test_ogr_capabilities.py).

But then I discovered [this GDAL PR](https://github.com/OSGeo/gdal/pull/3311#event-4137705427), so started investigating if `OGR ExportToJSON()`could be the cause. I added that to the test, using a dummy memory Feature Layer.  But still ordering was preserved. 

Still I suspected `ExportToJSON`, also as the Docker environment was the only one with GDAL >= 3.3 and where it failed in production. I put logging output before/after the `ExportToJSON`-call within the pygeoapi OGRProvider](https://github.com/geopython/pygeoapi/blob/crs-features-ogc/pygeoapi/provider/ogr.py#L523), and there it swapped, where it shouldn't! This happens after the Transform from EPSG:28992 to EPSG:4258, the correct order should be 53.319833312 6.854125905. This is since the OGC API Features Part 2 extension (CRS support) allows for other Projections encoded in GeoJSON (with a `Content-Crs` HTTP Header), so this was printed:

```
[2023-03-23T12:40:37Z] {/pygeoapi/pygeoapi/provider/ogr.py:524} INFO - coords before export to JSON: POINT (53.319833312 6.854125905)
[2023-03-23T12:40:37Z] {/pygeoapi/pygeoapi/provider/ogr.py:527} INFO - coords after export to JSON: [6.854125905, 53.319833312]
```

Maybe my tests succeed as I use an in-memory dummy layer to create Features. The next thing I I need to know: how I can suppress OGR swapping coordinates within `ExportToJSON`. Using the GDAL Version `int` this may be made conditionally.

rouault commented:
oh ok, well https://github.com/OSGeo/gdal/pull/3311 actually implements the GeoJSON spec  which mentions "The first two elements are longitude and latitude, or easting and northing, precisely in that order and using decimal numbers". So whatever the CRS definition says about that. You could likely work around the issue by doing geom.AssignSpatialReference(None). In the absence of a known CRS, the GeoJSON writer will not mess with axis order

MTachon commented:
It is great the the axis ordering issue is fixed!

I am now wondering whether we should keep the `source_crs_wkt`/`target_crs_wkt` part of the `CrsTransformSpec` dataclass. I first thought that WKT representations were the *safest* way to "pass around" CRS definitions. But we are now mainly dealing with CRS URIs, both in pygeoapi config and with the `crs`/`bbox-crs` query parameters. I think we could drop the support for importing/exporting CRSs from/to their WKT representations, especially since it may result in problems depending on WKT versions, and GDAL-PROJ version vs pyproj-PROJ version. We can simply use the `get_crs_from_uri` util function to load CRSs from `CrsTransformSpec` instances. What do you think?

justb4 commented on tests/test_ogr_shapefile_provider.py:
```diff
@@ -91,10 +91,12 @@ def config_shapefile_28992():
+def crs_transform_spec():
+    return CrsTransformSpec(
+        source_crs_uri='http://www.opengis.net/def/crs/OGC/1.3/CRS84',
+        source_crs_wkt=pyproj.CRS.from_epsg(4326),
```
For me it is not clear if `EPSG:4326` is http://www.opengis.net/def/crs/OGC/1.3/CRS84 or http://www.opengis.net/def/crs/EPSG/0/4326. But you are right, in light of current use, but on the other hand the Wkt's are not used within the `OGRProvider`. Also your next comment is of interest here.

MTachon replied:
For me, EPSG is a registry of spatial reference systems, geodetic datums, transformations, etc. and the EPSG codes are meant to identify these entities. I regard the CRS http://www.opengis.net/def/crs/OGC/1.3/CRS84 as not being part of this registry and not having an EPSG code identifying it there. The definition of a given CRS identified by an EPSG code specifies the axis-ordering of the CRS. http://www.opengis.net/def/crs/OGC/1.3/CRS84 is a URN from the OGC URNs namespace identifying a CRS. Appart from that, and not taking into account axis-ordering, http://www.opengis.net/def/crs/OGC/1.3/CRS84 and http://www.opengis.net/def/crs/EPSG/0/4326 are the exact same geographic CRSs as they are based on the same datum, have the same coordinate units etc.

MTachon commented:
Yes I understand, just thought that we could also use the `util.get_crs_from_uri` function to create `pyproj.CRS` instances (`util.get_crs_from_uri(crs_transform_spec.source_crs_uri)`/`util.get_crs_from_uri(crs_transform_spec.target_crs_uri)`), instead of using `pyproj.CRS.from_wkt(crs_transform_spec.source_crs_wkt)`/`pyproj.CRS.from_wkt(crs_transform_spec.target_crs_wkt)`, also in `PostgreSQLProvider` and in the `crs_transform` decorator.

About your suggestion of embedding transform functions in `CrsTransformSpec` instances, that could be a good idea. This is actually the approach I chose initially, when focusing only on support for  `PostgreSQLProvider`. I then passed a transform function instead of CRS definitions (WKT representations). Ended up passing around CRS definitions as they could also be used in other providers that would need their own transform functions. These transform functions could be made from these CRS definitions. I thought that using CRS definitions instead of functions may be better, as the `CrsTransformSpec` objects would be smaller/simpler, and they would require less maintenance/changes over time, if pygeoapi needs to support other types of providers that would require their own transform functions. But again, if these transform functions are mostly the same (one for `shapely` geometrical objects, and one for GeoJSON-like `dict` features), passing transform functions in `CrsTransformSpec` instances could remove a few lines of codes in the providers implementation. I do not think there is a right or wrong approach for that, we just need to chose what we prefer.

Yet another approach could be to create a `@singledispatch` generic function that takes the provider's type as parameter. This dispatch function would return the appropriate transform function, depending of the type of the provider. This can remove the creation-of-transform-function logic from the providers' implementation, and if several providers can use the same transform function, they just need to be registered accordingly, no code duplication there.

justb4 commented:
(The failing build is due to a remote WFS https://demo.mapserver.org/cgi-bin/wfs connection error, hopefully transient, as nothing was changed there, though I see connection issues locally as well with certificate issues).

@MTachon thanks for suggestions. Remarks:

* indeed it looks a bit overkill to first create a `pyproj.CRS` object from an URI and then convert to a WKT to again create a `pyproj.CRS` object for the `CrsTransformSpec`. Why not pass the CRS objects within `CrsTransformSpec` together with the URIs. But ok, this was before the whole GDAL issue. 
* as for embedding a Transform function(s) or facilitate Providers in other ways: yes, but see below.
* (new suggestion) the CRS-functionality is a bit spread-out (`util.py`, `api.py`, globals in both) . Maybe bundle in a single module `pygeoapi/crs.py`.

For now, there is a stable/working version backed by many unit tests. For sake of progress and to allow for easier additional development, I want to try to get this PR #1174 reviewed/merged and make the three above changes in additional issues/PR(s). They are enhancements, embellishments on the never-ending road to perfection :-). It is already for us quite some merge effort as we merge this branch into `crs-features-ogc-adr` as well. For all, like for you, it wil be easier to work from the `master` branch.

I hope @tomkralidis and/or @francbartoli may review. I will therefore remove the `[WIP]` prefix from the PR-title.

justb4 commented:
Strange, in cases the WFS responds with a certificate problem, but like in browser certificate seems valid:
```
ogrinfo -ro -al WFS:https://demo.mapserver.org/cgi-bin/wfs WFSLayerMetadata
ERROR 1: SSL certificate problem: certificate has expired
ERROR 1: Error returned by server : SSL certificate problem: certificate has expired (60)
```
Same error with
`curl -v  https://demo.mapserver.org/cgi-bin/wfs` but not on all systems.
LE cert seems ok, gets 'A' ranking on SSLLabs test...

justb4 commented:
With commit [fd16159](https://github.com/geopython/pygeoapi/pull/1174/commits/fd16159f94283a527e9e115adae38974604c346f) All OGC CITE tests for Part 1 and Part 2 are now succeeding!

![image](https://user-images.githubusercontent.com/582630/229136448-856ccc81-a39a-4194-a476-36fbf695de92.png)

tomkralidis commented on pygeoapi/provider/geojson.py:
```diff
@@ -130,6 +131,7 @@ def _load(self, skip_geometry=None, properties=[], select_properties=[]):
                                    if k in set(self.properties) | set(select_properties)}  # noqa
         return data
 
+    @crs_transform
```
Does this decorate mean an extra hop for all provider `query` and `get` functions?

MTachon replied:
The decorator iterates through the `features` returned by the providers' `query` and `get` functions, transform the geometries' coordinates if needed (meaning that `crs_transform_spec != None` in `query()` and `get()` calls in `api.py`) before the `Feature`/`FeatureCollection` is assigned to the `content` variable in `api.py`. The decorator approach was used to avoid code duplication (fewer lines of code if the transformation logic is taken out of the providers' implementation), for simplicity and for supporting coordinates transformation for all providers. `PostgreSQLProvider` has its own implementation and does not use the `@crs_transform` decorator. The implementation for the `PostgreSQL` provider is more optimized as it does not re-iterate through the returned `features`, and we may want to optimize the coordinates transformation implementation of other providers later, especially if these providers have *native support* for coordinates transformation. In practice, re-iterating through some thousands of features would not result in much overhead.

justb4 replied:
Yes, agree. Also the OGRProvider does its own transforms. Our plan is to first have #1174 merged, having a stable version, CITE and INSPIRE compliant. Then do refactoring in separate PRs:

- move all CRS-dedicated/related functionality and constants spread over `util.py` and `api.py`  into a new module `pygeoapi/crs.py`
- enhance `CrsTransformSpec` with direct-to-call transform functions as `partials` (e.g. for Shapely and GeoJSON dicts) that Providers without native CRS-support can call in their Feature-Collection loop
- upgrade Providers to using the CrsTransformSpec-based transforms, removing the Decorator, maybe even completely

GeoSander commented:
@justb4 Now that #1152 has been merged, please let me know if you need a hand resolving conflicts.

justb4 commented:
Thanks @GeoSander ! Yes, will be quite a merge from master, but seen these changes from #1152 before, IntelliJ will be my friend. I'll let know.

justb4 commented on docs/source/data-publishing/ogcapi-features.rst:
```diff
@@ -14,27 +14,35 @@ Providers
+
 .. csv-table::
-   :header: Provider, property filters/display, resulttype, bbox, datetime, sortby, skipGeometry, CQL, transactions
+   :header: Provider, property filters/display, resulttype, bbox, datetime, sortby, skipGeometry, CQL, transactions, crs
```
The 'crs' here refers to the parameter `crs` which is lowercase. You still want to change to 'CRS'?

justb4 commented on pygeoapi/util.py:
```diff
@@ -541,6 +581,209 @@ def get_envelope(coords_list: List[List[float]]) -> list:
+
+
+# Type for shapely geometrical objects.
+GeomObject = Union[
```
Done.

GeoSander commented:
FYI @justb4: I did a working tree diff with the current master and did not find anything wrong or missing (during the merge), so in that sense it looks good to me 👍

justb4 commented:
@GeoSander that is good to hear! Hopefully @tomkralidis finds some time to look at the last resolutions from the review.
